extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayMultiDatasourceBeanGenerator.java

```diff
@@ -0,0 +1,140 @@
+package io.quarkus.flyway;
+
+import java.util.Collection;
+import java.util.HashSet;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.Produces;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.sql.DataSource;
+
+import org.flywaydb.core.Flyway;
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationValue;
+import org.jboss.jandex.DotName;
+
+import io.quarkus.arc.deployment.GeneratedBeanBuildItem;
+import io.quarkus.arc.processor.DotNames;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.util.HashUtil;
+import io.quarkus.flyway.runtime.FlywayProducer;
+import io.quarkus.gizmo.BytecodeCreator;
+import io.quarkus.gizmo.ClassCreator;
+import io.quarkus.gizmo.FieldCreator;
+import io.quarkus.gizmo.FieldDescriptor;
+import io.quarkus.gizmo.MethodCreator;
+import io.quarkus.gizmo.MethodDescriptor;
+import io.quarkus.gizmo.ResultHandle;
+
+/**
+ * Generates the CDI producer bean for {@link Flyway} at built time.<br>
+ * Supports multiple {@link Named} {@link DataSource}s.
+ * <p>
+ * It {@link Produces} {@link Flyway} instances for every {@link Named} {@link DataSource}.<br>
+ * All {@link Flyway} instances get named the same way as the {@link DataSource}s,<br>
+ * appended by the postfix {@value #FLYWAY_BEANNAME_POSTFIX}.
+ * 
+ * @author JohT
+ */
+class FlywayMultiDatasourceBeanGenerator {
+
+    public static final String FLYWAY_BEANNAME_POSTFIX = "_flyway";
+
+    private static final String FLYWAY_PRODUCER_BEAN_NAME = "FlywayMultiDataSourceProducer";
+    private static final String FLYWAY_PRODUCER_PACKAGE_NAME = FlywayProducer.class.getPackage().getName();
+    private static final String FLYWAY_PRODUCER_TYPE_NAME = FLYWAY_PRODUCER_PACKAGE_NAME + "." + FLYWAY_PRODUCER_BEAN_NAME;
+
+    private static final String DEFAULT_DATASOURCE = "";
+    private static final int ACCESS_PACKAGE_PROTECTED = 0;
+
+    private final Collection<String> dataSourceNames = new HashSet<>();
+    private final BuildProducer<GeneratedBeanBuildItem> generatedBean;
+
+    public FlywayMultiDatasourceBeanGenerator(Collection<String> dataSourceNames,
```

Comment:
I don't think we should name things `MultiDatasource`. It's an implementation detail.

Reply:
ok